roles/lib_openshift/src/class/oc_configmap.py

```diff
@@ -0,0 +1,183 @@
+                 name,
+                 from_file,
+                 from_literal,
+                 state,
+                 namespace,
+                 kubeconfig='/etc/origin/master/admin.kubeconfig',
+                 verbose=False):
+        ''' Constructor for OpenshiftOC '''
+        super(OCConfigMap, self).__init__(namespace, kubeconfig=kubeconfig, verbose=verbose)
+        self.name = name
+        self.state = state
+        self._configmap = None
+        self._inc_configmap = None
+        self.from_file = from_file if from_file is not None else {}
+        self.from_literal = from_literal if from_literal is not None else {}
+
+    @property
+    def configmap(self):
+        if self._configmap is None:
+            self._configmap = self.get()
+
+        return self._configmap
+
+    @configmap.setter
+    def configmap(self, inc_map):
+        self._configmap = inc_map
+
+    @property
+    def inc_configmap(self):
+        if self._inc_configmap is None:
+            results = self.create(dryrun=True, output=True)
+            self._inc_configmap = results['results']
+
+        return self._inc_configmap
+
+    @inc_configmap.setter
+    def inc_configmap(self, inc_map):
+        self._inc_configmap = inc_map
+
+    def from_file_to_params(self):
+        '''return from_files in a string ready for cli'''
+        return ["--from-file={}={}".format(key, value) for key, value in self.from_file.items()]
+
+    def from_literal_to_params(self):
+        '''return from_literal in a string ready for cli'''
+        return ["--from-literal={}={}".format(key, value) for key, value in self.from_literal.items()]
+
+    def get(self):
+        '''return a configmap by name '''
+        results = self._get('configmap', self.name)
+        if results['returncode'] == 0 and results['results'][0]:
+            self.configmap = results['results'][0]
+
+        if results['returncode'] != 0 and '"{}" not found'.format(self.name) in results['stderr']:
+            results['returncode'] = 0
+
+        return results
+
+    def delete(self):
+        '''delete a configmap by name'''
+        return self._delete('configmap', self.name)
+
+    def create(self, dryrun=False, output=False):
+        '''Create a configmap
+
+           :dryrun: Product what you would have done. default: False
+           :output: Whether to parse output. default: False
+        '''
+
+        cmd = ['create', 'configmap', self.name]
+        if self.from_literal is not None:
+            cmd.extend(self.from_literal_to_params())
+
+        if self.from_file is not None:
+            cmd.extend(self.from_file_to_params())
+
+        if dryrun:
+            cmd.extend(['--dry-run', '-ojson'])
+
+        results = self.openshift_cmd(cmd, output=output)
+
+        return results
+
+    def update(self):
+        '''run update configmap '''
+        return self._replace_content('configmap', self.name, self.inc_configmap)
+
+    def needs_update(self):
```

Comment:
>  I imagine we will want to build in retry loops and timeouts.

Again, I think this really depends on what you want to do here -- if your intent is that you have the _complete_ and _correct_ version of state for an object, do not send the resource version and ensure that your version is what exists on the server. If you mean to _edit_ specific fields of an object, use `oc patch`.

> Can patch perform multiple edits at a single time?

Yes, you can for example do 
```
oc patch type/name --patch='{"spec":{"field": "value","otherfield": "whateverelse"}}"
```

>  Isn't this susceptible to the race conditions you mentioned above?

No. The race I spoke of is for changes to unrelated fields in the object.

> It seems to me that the situation you described would result in whoever made the edits last would win. This is kind of the same behavior as replace except we attempt to obey resource versions. 

Right, the person who makes the last edits does win. The critical difference is that with `patch` you say "I wan _this_ field to change" and thereby you honor edits others make to _unrelated_ fields. When you `oc replace` you say "I want the entire object to look like this".

>  As soon as patch makes changes and another call to patch occurs you run the risk of someone making a different change and now the module cannot, with confidence, return that the desired "state" was achieved.

This will always happen no matter the means by which objects are changed on the API server. This is intended. When multiple people edit in a sequence the last edit will be the one that is reflected in the current state of the object. That is not the racy behavior I mentioned above.
```